### DISCUSSION TITLE
Simulations

### DISCUSSION THREAD

**Discussion by kevinmsmith214** [Comet-Robotics/URC#15] [Autonomous]
How realistic does the simulation have to be and how can we build it?

**Comment by hectorv117** [comment]
I think it'll depend a lot on what the main purpose of the simulation is. If we just want to test out our control/nav2 nodes it should be pretty straight forward to simulate a rough geometric approximation of the 4 wheeled test bed with the expected physics (wheel measurements, angular rotation and velocities etc.). This way we can test out or ros2 code and see if we're able to move/control the simplified robot as expected. Once we actually have more cad files of the different parts we could transform these into .urdf files that gazeebo uses to simulate a more accurate physical representation of the rover. (We could also try to find and use other rover like urdf representations online for now).

As far as the autonomy side of things, since we'll be using the realsense d435 rgbd camera we would first need to find a way to simulate the camera/image data that it produces in ros2 within gazebo or possibly unity. We could then customize the sim environment with different obstacles to test out our obstacle avoidance algos. However, building outdoor maps in simulation using only visual/rgbd seems like it'll be pretty difficult as the outdoor env simulation wont have very many features or texture which is pretty important for visual odometry. If we were to use Unity maybe we could use a premade mars/desert like simulation env. Or maybe we could first build an outdoor map using the actual camera irl and find a way to store and use that map inside simulation? 

## Very roughly speaking, we would first need to:

### Basic ROS2 robot control and navigation simulation 

1. Build/use a simplified geometric and physical urdf (for gazeebo) representation of our robot (we could make the robot physics/measurements progressively more accurate)
2. Integrate the robot and env simulation with our ros2 code for testing 

### Simulated Autonomy

1. find a way to simulate the realsense d435 camera in ros2 (or something similar to the image data it produces)
2. find a way to build an outdoor desert like map and use it in simulation:
   - potentially use a premade mars like simulation environment in Unity?
   - try to build our own map irl carrying the camera around somewhere near utd
3. If we don't need a full blown SLAM built map we could still to test basic obstacle avoidance algos relatively easily in gazeebo or unity
